src/Compilers/CSharp/Test/Semantic/Semantics/GenericConstraintsTests.cs

```diff
@@ -2674,5 +2904,103 @@ public unsafe class C<U> where U : unmanaged
                 //     public void M1<T>() where T : T* { }
                 Diagnostic(ErrorCode.ERR_BadConstraintType, "T*").WithLocation(4, 35));
         }
+
+        [Fact]
+        public void UnmanagedConstraints_EnumWithUnmanaged()
+        {
+            Action<ModuleSymbol> validator = module =>
+            {
+                var typeParameter = module.GlobalNamespace.GetTypeMember("Test").TypeParameters.Single();
+
+                Assert.True(typeParameter.HasUnmanagedTypeConstraint);
+                Assert.True(typeParameter.HasValueTypeConstraint);
+                Assert.False(typeParameter.HasReferenceTypeConstraint);
+                Assert.False(typeParameter.HasConstructorConstraint);
+
+                Assert.Equal("Enum", typeParameter.ConstraintTypes().Single().Name);
+
+                Assert.True(typeParameter.IsValueType);
+                Assert.False(typeParameter.IsReferenceType);
+            };
+
+            CompileAndVerify(
+                source: "public class Test<T> where T : unmanaged, System.Enum {}",
+                sourceSymbolValidator: validator,
+                symbolValidator: validator);
+        }
+
+        [Fact]
+        public void UnmanagedConstraints_NestedInGenericType()
+        {
+            var code = @"
+public class Wrapper<T>
+{
+    public enum E
+    {
+    }
+
+    public struct S
+    {
+    }
+}
+public class Test
+{
+    void IsUnmanaged<T>() where T : unmanaged { }
+    void IsEnum<T>() where T : System.Enum { }
+    void IsStruct<T>() where T : struct { }
+    void IsNew<T>() where T : new() { }
+    
+
+    void User()
+    {
+        IsUnmanaged<Wrapper<int>.E>();
+        IsEnum<Wrapper<int>.E>();
+        IsStruct<Wrapper<int>.E>();
+        IsNew<Wrapper<int>.E>();
+
+        IsUnmanaged<Wrapper<int>.S>();          // Invalid
+        IsEnum<Wrapper<int>.S>();               // Invalid
```

Comment:
Why are these invalid?

Reply:
`IsEnum<Wrapper<int>.S>()` is invalid because `struct S` is not an `enum`. Did you mean to point to this test?

---
In reply to: [177175300](https://github.com/dotnet/roslyn/pull/25680#discussion_r177175300) [](ancestors = 177175300)

Reply:
We consider all types within a generic type managed (has nothing to do with this feature). For example, the following code is invalid:
`Wrapper<int>.S* ptr = null;`
error CS0208: Cannot take the address of, get the size of, or declare a pointer to a managed type ('Wrapper<int>.S')

---
In reply to: [177223126](https://github.com/dotnet/roslyn/pull/25680#discussion_r177223126) [](ancestors = 177223126)